### PR TITLE
Add API for create/edit/delete CDN

### PR DESCRIPTION
**Pull request by jifyang**
Add RESTful API for create/edit/delete CDN

### CODE REVIEW

File: docs/source/development/traffic_ops_api/v12/cdn.rst [hunk 1]
```diff
@@ -82,6 +82,143 @@ CDN
...
+
+  +------------------+--------+-------------------------------------------------+
+  |    Parameter     |  Type  |                   Description                   |
+  +==================+========+=================================================+
+  | ``name``         | string | CDN name.                                       |
+  +------------------+--------+-------------------------------------------------+
+  | ``id``           |  int   | CDN id.                                         |
+  +------------------+--------+-------------------------------------------------+
+
+  **Response Example** ::
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cdn.rst]: POST, PUT, PATCH and DELETE should all return an alerts node

https://github.com/Comcast/traffic_control/wiki/Traffic-Ops-API-Guidelines-v1.x#response-payload

Comment (jifyang) [inline docs/source/development/traffic_ops_api/v12/cdn.rst]: Accept. The code is updated.

File: traffic_ops/app/lib/API/Cdn.pm [hunk 2]
```diff
@@ -62,6 +62,121 @@ sub name {
...
+sub create {
+	my $self   = shift;
+	my $params = $self->req->json;
+
+	if ( !&is_oper($self) ) {
+		return $self->forbidden();
+	}
+
+	if ( !defined($params) ) {
+		return $self->alert("parameters should in json format.");
```
Comment (jheitz200) [inline traffic_ops/app/lib/API/Cdn.pm]: Please change this to `parameters must be in JSON format.`

File: traffic_ops/app/lib/API/Cdn.pm [hunk 3]
```diff
@@ -62,6 +62,121 @@ sub name {
...
+	}
+
+	if ( !defined($params->{name}) ) {
+		return $self->alert("parameter name is must.");
```
Comment (jheitz200) [inline traffic_ops/app/lib/API/Cdn.pm]: Please change this to `CDN 'name' is required.`

File: traffic_ops/app/lib/API/Cdn.pm [hunk 4]
```diff
@@ -62,6 +62,121 @@ sub name {
...
+	}
+
+	my $existing = $self->db->resultset('Cdn')->search( { name => $params->{name} } )->single();
+	if ( $existing ) {
+		$self->app->log->error( "cdn name conflict: " . $params->{name} );
```
Comment (jheitz200) [inline traffic_ops/app/lib/API/Cdn.pm]: Please change this to `$self->app->log->error( "a cdn with name ' . $params->{name} . ' already exists." );`

File: traffic_ops/app/lib/API/Cdn.pm [hunk 5]
```diff
@@ -62,6 +62,121 @@ sub name {
...
+	};
+	my $insert = $self->db->resultset('Cdn')->create($value);
+	$insert->insert();
+
+	my $rs = $self->db->resultset('Cdn')->find( { id => $insert->id } );
+	if ( defined($rs) ) {
+		my $response;
+		$response->{id}     = $rs->id;
+		$response->{name}   = $rs->name;
+		&log( $self, "create cdn " . $rs->name . ", id is " . $rs->id, "APICHANGE" );
```
Comment (jheitz200) [inline traffic_ops/app/lib/API/Cdn.pm]: Please change this to `"Created CDN with id: " . $rs->id . " and  name: " . $rs->name, "APICHANGE" );`

File: traffic_ops/app/lib/API/Cdn.pm [hunk 6]
```diff
@@ -62,6 +62,121 @@ sub name {
 	$self->success( \@data );
 }
 
+
+sub create {
+	my $self   = shift;
+	my $params = $self->req->json;
+
+	if ( !&is_oper($self) ) {
+		return $self->forbidden();
```
Comment (jheitz200) [inline traffic_ops/app/lib/API/Cdn.pm]: Please add an error message. `return $self->alert("You must be an ADMIN or OPER to perform this operation!");`

File: traffic_ops/app/lib/API/Cdn.pm [hunk 7]
```diff
@@ -62,6 +62,121 @@ sub name {
...
+	return $self->alert("create cdn failed.");
+}
+
+sub update {
+	my $self   = shift;
+	my $id     = $self->param('id');
+	my $params = $self->req->json;
+
+	if ( !&is_oper($self) ) {
+		return $self->forbidden();
```
Comment (jheitz200) [inline traffic_ops/app/lib/API/Cdn.pm]: Please add an error message. `return $self->alert("You must be an ADMIN or OPER to perform this operation!");`

File: traffic_ops/app/lib/API/Cdn.pm [hunk 8]
```diff
@@ -62,6 +62,121 @@ sub name {
...
+	}
+
+	my $cdn = $self->db->resultset('Cdn')->find( { id => $id } );
+	if ( !defined($cdn) ) {
+		return $self->not_found();
+	}
+
+	if ( !defined($params) ) {
+		return $self->alert("parameters should in json format.");
```
Comment (jheitz200) [inline traffic_ops/app/lib/API/Cdn.pm]: Please change this to `parameters must be in JSON format.`

File: traffic_ops/app/lib/API/Cdn.pm [hunk 9]
```diff
@@ -62,6 +62,121 @@ sub name {
...
+	}
+
+	if ( !defined($params->{name}) ) {
+		return $self->alert("parameter name is must.");
```
Comment (jheitz200) [inline traffic_ops/app/lib/API/Cdn.pm]: Please change this to `CDN 'name' is required.`

File: traffic_ops/app/lib/API/Cdn.pm [hunk 10]
```diff
@@ -62,6 +62,121 @@ sub name {
...
+	}
+
+	my $existing = $self->db->resultset('Cdn')->search( { name => $params->{name} } )->single();
+	if ( $existing && $existing->id != $cdn->id ) {
+		$self->app->log->error( "cdn name conflict: " . $params->{name} );
```
Comment (jheitz200) [inline traffic_ops/app/lib/API/Cdn.pm]: Please change this to `$self->app->log->error( "a cdn with name '" . $params->{name} . "' already exists." );`

File: traffic_ops/app/lib/API/Cdn.pm [hunk 11]
```diff
@@ -62,6 +62,121 @@ sub name {
...
+		name => $params->{name},
+	};
+	$cdn->update($value);
+
+	my $rs = $self->db->resultset('Cdn')->find( { id => $id } );
+	if ( defined($rs) ) {
+		my $response;
+		$response->{id}     = $rs->id;
+		$response->{name}   = $rs->name;
+		&log( $self, "update cdn " . $rs->id . " with name " . $rs->name, "APICHANGE" );
```
Comment (jheitz200) [inline traffic_ops/app/lib/API/Cdn.pm]: Please change this to `"Updated CDN name '" . $rs->name . "' for id: " . $rs->id, "APICHANGE" );`

File: docs/source/development/traffic_ops_api/v12/cdn.rst [hunk 12]
```diff
@@ -82,6 +82,172 @@ CDN
...
+  +-----------------+----------+------------------------------------------------+
+  |  Parameter      |  Type    |           Description                          |
+  +=================+==========+================================================+
+  |  ``alerts``     |  array   |  A collection of alert messages.               |
+  +-----------------+----------+------------------------------------------------+
+  |  ``>level``     |  string  |  success, info, warning or error.              |
+  +-----------------+----------+------------------------------------------------+
+  |  ``>text``      |  string  |  Alert message.                                |
+  +-----------------+----------+------------------------------------------------+
+  |  ``version``    |  string  |                                                |
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cdn.rst]: i don't think version applies anymore

File: traffic_ops/app/lib/API/Cdn.pm [hunk 13]
```diff
@@ -62,6 +62,121 @@ sub name {
 	$self->success( \@data );
 }
 
+
+sub create {
+	my $self   = shift;
+	my $params = $self->req->json;
+
```
Comment (jheitz200) [inline traffic_ops/app/lib/API/Cdn.pm]: this api should also include the field `dnssec_enabled`

File: traffic_ops/app/lib/API/Cdn.pm (hunk 2 above)
Comment (jifyang) [inline traffic_ops/app/lib/API/Cdn.pm]: Accept.

File: traffic_ops/app/lib/API/Cdn.pm (hunk 3 above)
Comment (jifyang) [inline traffic_ops/app/lib/API/Cdn.pm]: Accept.

File: traffic_ops/app/lib/API/Cdn.pm (hunk 4 above)
Comment (jifyang) [inline traffic_ops/app/lib/API/Cdn.pm]: Accept.

File: traffic_ops/app/lib/API/Cdn.pm (hunk 5 above)
Comment (jifyang) [inline traffic_ops/app/lib/API/Cdn.pm]: Accept.

File: traffic_ops/app/lib/API/Cdn.pm (hunk 8 above)
Comment (jifyang) [inline traffic_ops/app/lib/API/Cdn.pm]: Accept.

File: traffic_ops/app/lib/API/Cdn.pm (hunk 9 above)
Comment (jifyang) [inline traffic_ops/app/lib/API/Cdn.pm]: Accept.

File: traffic_ops/app/lib/API/Cdn.pm (hunk 10 above)
Comment (jifyang) [inline traffic_ops/app/lib/API/Cdn.pm]: Accept.

File: traffic_ops/app/lib/API/Cdn.pm (hunk 11 above)
Comment (jifyang) [inline traffic_ops/app/lib/API/Cdn.pm]: Accept.

File: docs/source/development/traffic_ops_api/v12/cdn.rst (hunk 12 above)
Comment (jifyang) [inline docs/source/development/traffic_ops_api/v12/cdn.rst]: Accept.

File: traffic_ops/app/lib/API/Cdn.pm (hunk 13 above)
Comment (jifyang) [inline traffic_ops/app/lib/API/Cdn.pm]: Accept. The code is updated.

**Pull request merged**